internal/cmd/lsp-record/lsp-record.go

```diff
@@ -0,0 +1,349 @@
+// Command lsp-record provides an lsp proxy which records the session. It is intended
+package main
+
+import (
+	"archive/zip"
+	"bufio"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"os/exec"
+	"os/signal"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sourcegraph/go-langserver/pkg/lsp"
+	"github.com/sourcegraph/jsonrpc2"
+)
+
+// Request is the subset of a JSONRPC2 request payload we want to record
+type Request struct {
+	Method string           `json:"method"`
+	Params *json.RawMessage `json:"params,omitempty"`
+}
+
+func encode(w io.Writer, v interface{}) error {
+	b, err := json.Marshal(v)
+	if err != nil {
+		return err
+	}
+	if _, err := w.Write(b); err != nil {
+		return err
+	}
+	if _, err := w.Write([]byte{'\n'}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func writeJSONRPC2Requests(r io.Reader, w io.Writer) error {
+	stream := bufio.NewReader(r)
+	codec := jsonrpc2.VSCodeObjectCodec{}
+
+	for {
+		var req Request
+		if err := codec.ReadObject(stream, &req); err != nil {
+			return err
+		}
+		if req.Method != "" {
+			err := encode(w, req)
+			if err != nil {
+				return err
+			}
+		}
+	}
+}
+
+func retryDial(network, address string) (net.Conn, error) {
+	conn, err := net.DialTimeout(network, address, time.Second)
+	for i := 0; err != nil && i < 5; i++ {
+		time.Sleep(time.Second)
+		conn, err = net.DialTimeout(network, address, time.Second)
+	}
+	return conn, err
+}
+
+func massageGitHubArchive(r *zip.ReadCloser) {
+	// Strip out top-level dir so the root is the root of the repo
+	for i, file := range r.File {
+		r.File[i].Name = file.Name[strings.Index(file.Name, "/"):]
+	}
+}
+
+func openArchive(originalRootURI string) (*zip.ReadCloser, error) {
```

Comment:
It took me a second to realize what this function was doing. Could this function be renamed to something like `fetchArchiveForRepo` and/or change the arguments to be `(repo, rev)? Just to improve clarity